tokio/src/sync/once_cell.rs

```diff
@@ -399,6 +401,44 @@ impl<T> OnceCell<T> {
     pub fn take(&mut self) -> Option<T> {
         std::mem::take(self).into_inner()
     }
+
+    #[inline(always)]
+    /// Returns a future that resolves once a value has been written to the `OnceCell`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tokio::sync::OnceCell;
+    ///
+    /// static LIFE: OnceCell<u8> = OnceCell::const_new();
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     tokio::task::spawn(async {
+    ///         let life: &u8 = LIFE.wait().await;
+    ///         assert_eq!(*life, 42);
+    ///     });
+    ///
+    ///     assert!(LIFE.set(42).is_ok());
+    /// }
+    /// ```
+    pub fn wait(&self) -> impl Future<Output = &T> {
```

Comment:
I really don't like the approach of having a method for this and also having the type become a future. This should return a distinct custom future which is woken when the cell is initialized.